crates/wasmi/src/store.rs

```diff
@@ -977,6 +1011,36 @@ impl<T> Store<T> {
             .get(entity_index)
             .unwrap_or_else(|| panic!("failed to resolve stored host function: {entity_index:?}"))
     }
+
+    /// Sets a callback function that is executed whenever a WebAssembly
+    /// function is called from the host or a host function is called from
+    /// WebAssembly, or these functions return.
+    ///
+    /// The function is passed a `&mut T` to the underlying store, and a
+    /// [`CallHook`]. [`CallHook`] can be used to find out what kind of function
+    /// is being called or returned from.
+    ///
+    /// The callback can either return `Ok(())` or an `Err` with an
+    /// [`Error`]. If an error is returned, it is returned to the host
+    /// caller. If there are nested calls, only the most recent host caller
+    /// receives the error and it is not propagated further automatically. The
+    /// hook may be invoked again as new functions are called and returned from.
+    pub fn call_hook(
+        &mut self,
+        hook: impl FnMut(&mut T, CallHook) -> Result<(), Error> + Send + Sync + 'static,
+    ) {
+        self.call_hook = Some(CallHookWrapper(Box::new(hook)));
+    }
+
+    /// Executes the callback set by [`Store::call_hook`] if any has been set
+    /// and returns the value returned by the callback. If no callback has been
+    /// set, `Ok(())` is returned.
+    pub(crate) fn invoke_call_hook(&mut self, call_type: CallHook) -> Result<(), Error> {
+        match &mut self.call_hook {
+            None => Ok(()),
+            Some(call_hook) => call_hook.0(&mut self.data, call_type),
+        }
+    }
```

Comment:
Please see my non-review comment for improvements (optimizations) here. Also could you please see how the suggested code performs on your machine? I only have an M2 macbook to run benchmarks on.